test_circuit.md

```diff
@@ -58,52 +78,50 @@ impl<
         1 << 11
     }
 }
-
-// Now let's use the Circuit we've just implemented!
-fn main()-> Result<(), Error> {
-    let pp: PublicParameters<Bls12_381> = KZG10::<Bls12_381,DensePolynomial<BlsScalar>,>::setup(
-          1 << 12, false, &mut OsRng
-    )?;
-    // Initialize the circuit
-    let mut circuit: TestCircuit::<
-        Bls12_381,
-        JubjubParameters,
-    > = TestCircuit::default();
-    // Compile the circuit
-    let (pk, vd) = circuit.compile(&pp).unwrap();
-    // Generator
-    let (x, y) = JubJubParameters::AFFINE_GENERATOR_COEFFS;
-    let generator = JubJubAffine::new(x, y);
-    let point_f_pi: JubJubAffine = AffineCurve::mul(
-      &generator,
-      JubJubScalar::from(2u64).into_repr(),
-    ).into_affine();
-    let proof = {
-        let mut circuit = TestCircuit {
-            a: BlsScalar::from(20u64),
-            b: BlsScalar::from(5u64),
-            c: BlsScalar::from(25u64),
-            d: BlsScalar::from(100u64),
-            e: JubJubScalar::from(2u64),
-            f: point_f_pi,
-        };
-        circuit.gen_proof(&pp, pk, b"Test").unwrap()
+// Generate CRS
+let pp = KZG10::<Bls12_381, DensePolynomial<BlsScalar>>::setup(
+    1 << 12,
+    false,
+    &mut OsRng,
+)?;
+let mut circuit = TestCircuit::<Bls12_381, JubJubParameters>::default();
+// Compile the circuit
+let (pk_p, verifier_data) = circuit.compile(&pp)?;
+let (x, y) = JubJubParameters::AFFINE_GENERATOR_COEFFS;
+let generator: GroupAffine<JubJubParameters> = GroupAffine::new(x, y);
+let point_f_pi: GroupAffine<JubJubParameters> = AffineCurve::mul(
+    &generator,
+    JubJubScalar::from(2u64).into_repr(),
+)
+.into_affine();
+// Prover POV
+let proof = {
+    let mut circuit: TestCircuit<Bls12_381, JubJubParameters> = TestCircuit {
+        a: BlsScalar::from(20u64),
+        b: BlsScalar::from(5u64),
+        c: BlsScalar::from(25u64),
+        d: BlsScalar::from(100u64),
+        e: JubJubScalar::from(2u64),
+        f: point_f_pi,
     };
+    circuit.gen_proof(&pp, pk_p, b"Test")?
```

Comment:
no need to unwrap? also why use the question mark here?

Reply:
The question mark returns control flow to the caller with an error condition, in this case it will exit the program.

Reply:
Thank you, that one I figured out. What about unwrapping?

Reply:
Not sure what you mean by unwrapping here. If you mean calling `unwrap` instead of `?`, I would say that `?` is more correct here, since we want to return instead of panic. It's also more efficient and more succinct.

Reply:
Yes I meant calling unwrap. I didn't know ```?``` is more succinct. Good to know, thank you